OpenRA.Game/Map/MapCache.cs

```diff
@@ -378,4 +401,118 @@ public void Dispose()
 			});
 		}
 	}
+
+	public class MapDirectoryTracker
+	{
+		readonly FileSystemWatcher watcher;
+		readonly ModData modData;
+		readonly MapGrid mapGrid;
+		readonly IReadOnlyPackage package;
+		readonly MapClassification classification;
+
+		enum MapAction { Add, Delete, Update }
+		readonly Dictionary<string, MapAction> mapActionQueue = new Dictionary<string, MapAction>();
+
+		public MapDirectoryTracker(MapGrid mapGrid, IReadOnlyPackage package, MapClassification classification)
+		{
+			this.mapGrid = mapGrid;
+			this.package = package;
+			this.classification = classification;
+			modData = Game.ModData;
+
+			watcher = new FileSystemWatcher(package.Name);
+			watcher.Changed += (object sender, FileSystemEventArgs e) =>
+			{
+				if (e.ChangeType != WatcherChangeTypes.Changed)
+				{
+					return;
+				}
+
+				AddActionSub(MapAction.Update, e.FullPath);
+			};
+			watcher.Created += (object sender, FileSystemEventArgs e) => AddActionSub(MapAction.Add, e.FullPath);
+			watcher.Deleted += (object sender, FileSystemEventArgs e) => AddActionSub(MapAction.Delete, e.FullPath);
+			watcher.Renamed += (object sender, RenamedEventArgs e) => AddActionSub(MapAction.Add, e.FullPath, e.OldFullPath);
+
+			watcher.IncludeSubdirectories = true;
+			watcher.EnableRaisingEvents = true;
+		}
+
+		void AddActionSub(MapAction mapAction, string fullpath, string oldFullPath = "")
+		{
+			// ignore hidden files
+			if (mapAction != MapAction.Delete && (new FileInfo(fullpath).Attributes & FileAttributes.Hidden) == FileAttributes.Hidden)
+				return;
```

Comment:
I ignore hidden files here. But this method marks deleted maps as hidden. With the way I handled it, hidden deleted files still go through. They don't cause any troubles though

Reply:
What is the reason for ignoring hidden files?

Reply:
Tbf no reason, the engine already deals with garbage files in map directories. I just didn't like to constantly see hidden osx and zipping files when debugging